### PR TITLE
[MXNET-386] API Docs Generation

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
API docs generation (Write file). Create a file (an abstract class) that can be inherited to the API classes. 
@nswamy @yzhliu 
### Essentials ###
- [x] API file generator
- [x] Maven build integration
- [x] Current API integration
- [x] Operational and test checking
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+  def main(args: Array[String]) : Unit = {
+    val FILE_PATH = args(0)
+    absClassGen(FILE_PATH, true)
+    absClassGen(FILE_PATH, false)
+  }
+
+  def absClassGen(FILE_PATH : String, isSymbol : Boolean) : Unit = {
+    // scalastyle:off
+    val traitFunctions = initSymbolModule(isSymbol)
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: traitFunctions->classMethods
same everywhere

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    val traitfuncs = traitFunctions.filterNot(_.name.startsWith("_")).map(traitfunction => {
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: can we have these filtering condition in one place, may be declare a constant string somewhere...its in too many places across files and easy to miss if we refactor

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    val absClassDef = s"abstract class $packageName"
+    val finalStr = s"$apacheLicence\n$scalaStyle\n$packageDef\n$absClassDef {\n${traitfuncs.mkString("\n")}\n}"
+    import java.io._
+    val pw = new PrintWriter(new File(FILE_PATH + s"$packageName.scala"))
+    pw.write(finalStr)
+    pw.close()
+  }
+
+  // Generate ScalaDoc type
+  def ScalaDocGen(traitFunc : traitFunction) : String = {
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: I would give it a different name, sounds like inbuilt scala method, may be generateAPIDocFromBackend
also we follow camelCase for methodNames

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    } else {
+      throw new IllegalArgumentException(
+        s"Unrecognized arg field: $argType, ${commaRemoved.length}")
+    }
+
+  }
+
+
+  // List and add all the atomic symbol functions to current module.
+  private def initSymbolModule(isSymbol : Boolean): List[traitFunction] = {
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: initSymbolModule->getSymbolNDArrayMethods

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                case "type" => "typeOf"
+                case _ => traitarg.argName
+      }
+      s"  * @param $currArgName\t\t${traitarg.argDesc}"
+    })
+    val returnType = s"  * @return ${traitFunc.returnType}"
+    s"  /**\n${desc.mkString("\n")}\n${params.mkString("\n")}\n$returnType\n  */"
+  }
+
+  def defBodyGen(traitFunc : traitFunction, isSymbol : Boolean) : String = {
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: defBodyGen-> generateAPISignature

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    if (isSymbol) {
+      argDef += "name : String = null"
+      argDef += "attr : Map[String, String] = null"
+    }
+    s"def ${traitFunc.name} (${argDef.mkString(", ")}) : ${traitFunc.returnType}"
+  }
+
+
+  // Convert C++ Types to Scala Types
+  def typeConversion(in : String, argType : String = "", returnType : String) : String = {
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: please use this in the NDArrayAPI/SymbolAPI class and remove the redundant code

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.mxnet
+
+import org.apache.mxnet.init.Base._
+
+import scala.collection.mutable.ListBuffer
+
+private[mxnet] object APIDocGenerator{
```
[the remark this answers is not in the document]
Comment (lanking520) [reply scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: Does it seemed too long?

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  case class traitArg(argName : String, argType : String, argDesc : String, isOptional : Boolean)
```
[the remark this answers is not in the document]
Comment (lanking520) [reply scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: Done the change

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala (hunk 2 above)
Comment (lanking520) [reply scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: Let's put them all together in one class, will create a Class for that

**Comment by lanking520** [comment]
Add 1 line change in Sync with this PR: https://github.com/apache/incubator-mxnet/pull/11021

**Pull request merged**